### PR TITLE
Collision/Explosion optimisations

### PR DESCRIPTION
**Pull request by Spottedleaf**
See the commit notes:
https://github.com/PaperMC/Folia/commit/c6950e2d08a047975930bd9a706f31b085707b9b
https://github.com/PaperMC/Folia/commit/3b16d8fc1925ad51958346dfb9e5b786d5bc3648

Please report any issues with the branch on this PR.

### CODE REVIEW

File: patches/server/0029-Actually-optimise-explosions.patch [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
++                    );
++
++                    if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
++                        ++missedRays;
++                    }
++                }
++            }
++        }
++
++        return (double)missedRays / (double)totalRays;
```
Comment (Samsuik) [inline patches/server/0029-Actually-optimise-explosions.patch:305]:
```suggestion
+        return (float)missedRays / (float)totalRays;
```
This is different compared to vanilla `getSeenPercent` and could cause some issues with TNT cannons.

**Comment by Samsuik** [comment]
I have a concern that this pr could regress TNT cannon performance because of the removal of the cache in the optimize explosions patch. The TNT cannons used on Factions have large amounts of TNT and Sand upwards into the thousands in the same location. Cannon and Faction servers often disable random TNT spawn velocity as well to better utilise the cache when compressing TNT together.

I also had this exception
```
java.lang.NullPointerException: Cannot invoke "net.minecraft.world.level.block.state.IBlockData.emptyCollisionShape()" because "blockState" is null
	at net.minecraft.world.level.Explosion.clipsAnything(Explosion.java:267) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at net.minecraft.world.level.Explosion.getSeenFraction(Explosion.java:353) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at net.minecraft.world.level.Explosion.explode(Explosion.java:552) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at net.minecraft.world.level.Level.explode(Level.java:1494) ~[app:?]
	at net.minecraft.server.level.ServerLevel.explode(ServerLevel.java:1965) ~[app:?]
	at net.minecraft.world.level.Level.explode(Level.java:1460) ~[app:?]
	at net.minecraft.world.entity.item.PrimedTnt.explode(PrimedTnt.java:130) ~[app:?]
	at net.minecraft.world.entity.item.PrimedTnt.tick(PrimedTnt.java:90) ~[app:?]
	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1398) ~[app:?]
	at net.minecraft.world.level.Level.guardEntityTick(Level.java:1412) ~[app:?]
	at net.minecraft.server.level.ServerLevel.lambda$tick$12(ServerLevel.java:875) ~[app:?]
	at io.papermc.paper.threadedregions.RegionizedWorldData.forEachTickingEntity(RegionizedWorldData.java:543) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:854) ~[app:?]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1675) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1532) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

**Comment by Spottedleaf** [comment]
> I have a concern that this pr could regress TNT cannon performance because of the removal of the cache in the optimize explosions patch. The TNT cannons used on Factions have large amounts of TNT and Sand upwards into the thousands in the same location. Cannon and Faction servers often disable random TNT spawn velocity as well to better utilise the cache when compressing TNT together.
> 
> I also had this exception
> 
> ```
> java.lang.NullPointerException: Cannot invoke "net.minecraft.world.level.block.state.IBlockData.emptyCollisionShape()" because "blockState" is null
> 	at net.minecraft.world.level.Explosion.clipsAnything(Explosion.java:267) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.world.level.Explosion.getSeenFraction(Explosion.java:353) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.world.level.Explosion.explode(Explosion.java:552) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.world.level.Level.explode(Level.java:1494) ~[app:?]
> 	at net.minecraft.server.level.ServerLevel.explode(ServerLevel.java:1965) ~[app:?]
> 	at net.minecraft.world.level.Level.explode(Level.java:1460) ~[app:?]
> 	at net.minecraft.world.entity.item.PrimedTnt.explode(PrimedTnt.java:130) ~[app:?]
> 	at net.minecraft.world.entity.item.PrimedTnt.tick(PrimedTnt.java:90) ~[app:?]
> 	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1398) ~[app:?]
> 	at net.minecraft.world.level.Level.guardEntityTick(Level.java:1412) ~[app:?]
> 	at net.minecraft.server.level.ServerLevel.lambda$tick$12(ServerLevel.java:875) ~[app:?]
> 	at io.papermc.paper.threadedregions.RegionizedWorldData.forEachTickingEntity(RegionizedWorldData.java:543) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:854) ~[app:?]
> 	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1675) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1532) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at java.lang.Thread.run(Thread.java:833) ~[?:?]
> ```

I honestly didn't think that anyone running 1.20 would also be running factions. Although, Folia has the tnt optimisation patch for that disabled anyways (since day 1) due its usage of global state. Were you running Folia before and did you notice any problems with the tnt performance?

File: patches/server/0029-Actually-optimise-explosions.patch (hunk 1 above)
Comment (Spottedleaf) [reply patches/server/0029-Actually-optimise-explosions.patch]: changed to adhere back to vanilla

**Comment by Spottedleaf** [comment]
> I have a concern that this pr could regress TNT cannon performance because of the removal of the cache in the optimize explosions patch. The TNT cannons used on Factions have large amounts of TNT and Sand upwards into the thousands in the same location. Cannon and Faction servers often disable random TNT spawn velocity as well to better utilise the cache when compressing TNT together.
> 
> I also had this exception
> 
> ```
> java.lang.NullPointerException: Cannot invoke "net.minecraft.world.level.block.state.IBlockData.emptyCollisionShape()" because "blockState" is null
> 	at net.minecraft.world.level.Explosion.clipsAnything(Explosion.java:267) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.world.level.Explosion.getSeenFraction(Explosion.java:353) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.world.level.Explosion.explode(Explosion.java:552) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.world.level.Level.explode(Level.java:1494) ~[app:?]
> 	at net.minecraft.server.level.ServerLevel.explode(ServerLevel.java:1965) ~[app:?]
> 	at net.minecraft.world.level.Level.explode(Level.java:1460) ~[app:?]
> 	at net.minecraft.world.entity.item.PrimedTnt.explode(PrimedTnt.java:130) ~[app:?]
> 	at net.minecraft.world.entity.item.PrimedTnt.tick(PrimedTnt.java:90) ~[app:?]
> 	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1398) ~[app:?]
> 	at net.minecraft.world.level.Level.guardEntityTick(Level.java:1412) ~[app:?]
> 	at net.minecraft.server.level.ServerLevel.lambda$tick$12(ServerLevel.java:875) ~[app:?]
> 	at io.papermc.paper.threadedregions.RegionizedWorldData.forEachTickingEntity(RegionizedWorldData.java:543) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:854) ~[app:?]
> 	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1675) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1532) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> 	at java.lang.Thread.run(Thread.java:833) ~[?:?]
> ```

Resolved the exception https://github.com/PaperMC/Folia/pull/153/commits/259f2dacde0244d5b9f6876a186d0d294b6d9e6a

**Comment by Samsuik** [comment]
> > I have a concern that this pr could regress TNT cannon performance because of the removal of the cache in the optimize explosions patch. The TNT cannons used on Factions have large amounts of TNT and Sand upwards into the thousands in the same location. Cannon and Faction servers often disable random TNT spawn velocity as well to better utilise the cache when compressing TNT together.
> > I also had this exception
> > ```
> > java.lang.NullPointerException: Cannot invoke "net.minecraft.world.level.block.state.IBlockData.emptyCollisionShape()" because "blockState" is null
> > 	at net.minecraft.world.level.Explosion.clipsAnything(Explosion.java:267) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at net.minecraft.world.level.Explosion.getSeenFraction(Explosion.java:353) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at net.minecraft.world.level.Explosion.explode(Explosion.java:552) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at net.minecraft.world.level.Level.explode(Level.java:1494) ~[app:?]
> > 	at net.minecraft.server.level.ServerLevel.explode(ServerLevel.java:1965) ~[app:?]
> > 	at net.minecraft.world.level.Level.explode(Level.java:1460) ~[app:?]
> > 	at net.minecraft.world.entity.item.PrimedTnt.explode(PrimedTnt.java:130) ~[app:?]
> > 	at net.minecraft.world.entity.item.PrimedTnt.tick(PrimedTnt.java:90) ~[app:?]
> > 	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1398) ~[app:?]
> > 	at net.minecraft.world.level.Level.guardEntityTick(Level.java:1412) ~[app:?]
> > 	at net.minecraft.server.level.ServerLevel.lambda$tick$12(ServerLevel.java:875) ~[app:?]
> > 	at io.papermc.paper.threadedregions.RegionizedWorldData.forEachTickingEntity(RegionizedWorldData.java:543) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:854) ~[app:?]
> > 	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1675) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1532) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > 	at java.lang.Thread.run(Thread.java:833) ~[?:?]
> > ```
> 
> I honestly didn't think that anyone running 1.20 would also be running factions. Although, Folia has the tnt optimisation patch for that disabled anyways (since day 1) due its usage of global state. Were you running Folia before and did you notice any problems with the tnt performance?

I just help run a 1.20 cannon server running our own fork of Paper with tnt optimisations. There are quite a few modern factions servers out there with the biggest currently with 141 players at the time I am writing this and some of the older community is moving on trying to update. The biggest issue we face is that any tnt optimisations other than what is in Paper is paywalled and often break behaviour which stunts a lot of growth for new servers. In my experience both Paper and Folia have performance issues with TNT cannons.

**Comment by Yomamaeatstoes** [comment]
> > > I have a concern that this pr could regress TNT cannon performance because of the removal of the cache in the optimize explosions patch. The TNT cannons used on Factions have large amounts of TNT and Sand upwards into the thousands in the same location. Cannon and Faction servers often disable random TNT spawn velocity as well to better utilise the cache when compressing TNT together.
> > > I also had this exception
> > > ```
> > > java.lang.NullPointerException: Cannot invoke "net.minecraft.world.level.block.state.IBlockData.emptyCollisionShape()" because "blockState" is null
> > > 	at net.minecraft.world.level.Explosion.clipsAnything(Explosion.java:267) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at net.minecraft.world.level.Explosion.getSeenFraction(Explosion.java:353) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at net.minecraft.world.level.Explosion.explode(Explosion.java:552) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at net.minecraft.world.level.Level.explode(Level.java:1494) ~[app:?]
> > > 	at net.minecraft.server.level.ServerLevel.explode(ServerLevel.java:1965) ~[app:?]
> > > 	at net.minecraft.world.level.Level.explode(Level.java:1460) ~[app:?]
> > > 	at net.minecraft.world.entity.item.PrimedTnt.explode(PrimedTnt.java:130) ~[app:?]
> > > 	at net.minecraft.world.entity.item.PrimedTnt.tick(PrimedTnt.java:90) ~[app:?]
> > > 	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1398) ~[app:?]
> > > 	at net.minecraft.world.level.Level.guardEntityTick(Level.java:1412) ~[app:?]
> > > 	at net.minecraft.server.level.ServerLevel.lambda$tick$12(ServerLevel.java:875) ~[app:?]
> > > 	at io.papermc.paper.threadedregions.RegionizedWorldData.forEachTickingEntity(RegionizedWorldData.java:543) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:854) ~[app:?]
> > > 	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1675) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1532) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at io.papermc.paper.threadedregions.TickRegions$ConcreteRegionTickHandle.tickRegion(TickRegions.java:360) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at io.papermc.paper.threadedregions.TickRegionScheduler$RegionScheduleHandle.runTick(TickRegionScheduler.java:385) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at ca.spottedleaf.concurrentutil.scheduler.SchedulerThreadPool$TickThreadRunner.run(SchedulerThreadPool.java:525) ~[folia-server-1.20.1-R0.1-SNAPSHOT-reobf.jar:git-Folia-"3b16d8f"]
> > > 	at java.lang.Thread.run(Thread.java:833) ~[?:?]
> > > ```
> > 
> > 
> > I honestly didn't think that anyone running 1.20 would also be running factions. Although, Folia has the tnt optimisation patch for that disabled anyways (since day 1) due its usage of global state. Were you running Folia before and did you notice any problems with the tnt performance?
> 
> I just help run a 1.20 cannon server running our own fork of Paper with tnt optimisations. There are quite a few modern factions servers out there with the biggest currently with 141 players at the time I am writing this and some of the older community is moving on trying to update. The biggest issue we face is that any tnt optimisations other than what is in Paper is paywalled and often break behaviour which stunts a lot of growth for new servers. In my experience both Paper and Folia have performance issues with TNT cannons.

I did end up testing your claims @Samsuik about it effecting TNT Cannons, and it seems to not be the case. 
A video is here if you want to see it. ( https://youtu.be/SVdl3NBXATM ) It seems to be fine. 

I have been running on this branch to see if I would encounter any bugs, and it seems players occasionally die by fall damage when they teleport around. This happens rarely, but it does happen. Here are the plugins it has happened on

EpicHomes ( )

StormerTPAReloaded ( https://github.com/Loving11ish/StormerTPAReloaded )

StormerWarpsReloaded ( https://github.com/Loving11ish/StormerWarpsReloaded )

BetterRTP ( https://github.com/RonanPlugins/BetterRTP ) Even rarer.

Here is a video of it now ( https://youtu.be/sRbGOVQYFzM ) I think he was using a bad video recorder or something because it is like 6 fps.

**Comment by Samsuik** [comment]
> I did end up testing your claims @Samsuik about it effecting TNT Cannons, and it seems to not be the case. A video is here if you want to see it. ( https://youtu.be/SVdl3NBXATM ) It seems to be fine.

I started building a cannon specifically for testing and I finished last night. Here are the results of my crude testing:

```
Paper (optimize-explosions): 8.39 tps
Folia (this pr/optimisations branch): 3.94 region tps
Folia: 1.84 region tps
```

If you would like to test the performance of the cannon I made I uploaded the cannon to google drive.

Link: https://drive.google.com/file/d/1AKTyQekysAk5oA8xRVmq45GXgTDPTyp1/view?usp=sharing

**Comment by Yomamaeatstoes** [comment]
> > I did end up testing your claims @Samsuik about it effecting TNT Cannons, and it seems to not be the case. A video is here if you want to see it. ( https://youtu.be/SVdl3NBXATM ) It seems to be fine.
> 
> I started building a cannon specifically for testing and I finished last night. Here are the results of my crude testing:
> 
> ```
> Paper (optimize-explosions): 8.39 tps
> Folia (this pr/optimisations branch): 3.94 region tps
> Folia: 1.84 region tps
> ```
> 
> If you would like to test the performance of the cannon I made I uploaded the cannon to google drive.
> 
> Link: https://drive.google.com/file/d/1AKTyQekysAk5oA8xRVmq45GXgTDPTyp1/view?usp=sharing

That looks like something I would have to past using FAWE.

Would you be able to make a video, and perhaps a spark report of it? I would be interested to see the results.

**Comment by DarkyG2** [comment]
After some deep testing I can see a x3-5 more performance on a instant heavy load with huge faction cannons (1000 tnt at a time)  and after the first shot it stays at 20 tps, tested on a Intel® Core™ i9-13900 24 Core 64 GB DDR5 RAM compared to the old folia I get around 10-15 tps casually. (some paper config had to be changed in order to have all the tnt work) 

Can't wait for chunk (un)load to get fixed !

**Comment by Spottedleaf** [comment]
Merged to paper 1.20.2



> > I did end up testing your claims @Samsuik about it effecting TNT Cannons, and it seems to not be the case. A video is here if you want to see it. ( https://youtu.be/SVdl3NBXATM ) It seems to be fine.
> 
> I started building a cannon specifically for testing and I finished last night. Here are the results of my crude testing:
> 
> ```
> Paper (optimize-explosions): 8.39 tps
> Folia (this pr/optimisations branch): 3.94 region tps
> Folia: 1.84 region tps
> ```
> 
> If you would like to test the performance of the cannon I made I uploaded the cannon to google drive.
> 
> Link: https://drive.google.com/file/d/1AKTyQekysAk5oA8xRVmq45GXgTDPTyp1/view?usp=sharing

I've not tested this myself, however the upcoming paper 1.20.2 build will have the legacy optimize explosions and the performance from this branch.

**Comment by AltronMaxX** [comment]
Unable to redstone like on screenshot. Same problem with back side of piston
IMG_20230926_104526_000

**Comment by Spottedleaf** [comment]
> @Spottedleaf Unable to place redstone like on screenshot. Same problem with back side of piston IMG_20230926_104526_000

Resolved by https://github.com/PaperMC/Paper/commit/c207429b2135275a2c272b9932a2d8305329dd2f, will not be ported to this branch

**Pull request closed without merging**